azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -736,32 +731,159 @@ def do_retry():
 ]
 
 
+class ReconnectState(object):
+    """
+    Class which holds reconenct states as class variables.  Created to make code that reads like an enum without using an enum.
+
+    NEVER_CONNECTED: Ttransport has never been conencted.  This state is necessary because some errors might be fatal or transient,
+    depending on wether the transport has been connceted.  For example, a failed conenction is a transient error if we've connected
+    before, but it's fatal if we've never conencted.
+
+    WAITING_TO_RECONNECT: This stage is in a waiting period before reconnecting.
+
+    CONNECTED_OR_DISCONNECTED: The transport is either connected or disconencted.  This stage doesn't really care which one, so
+    it doesn't keep track.
+    """
+
+    NEVER_CONNECTED = "NEVER_CONNECTED"
+    WAITING_TO_RECONNECT = "WAITING_TO_RECONNECT"
+    CONNECTED_OR_DISCONNECTED = "CONNECTED_OR_DISCONNECTED"
+
+
 class ReconnectStage(PipelineStage):
     def __init__(self):
         super(ReconnectStage, self).__init__()
         self.reconnect_timer = None
-        self.virtually_connected = False
+        self.state = ReconnectState.NEVER_CONNECTED
         # connect delay is hardcoded for now.  Later, this comes from a retry policy
         self.reconnect_delay = 10
+        self.waiting_connect_ops = []
 
     @pipeline_thread.runs_on_pipeline_thread
     def _run_op(self, op):
         if isinstance(op, pipeline_ops_base.ConnectOperation):
-            self.virtually_connected = True
-            self.send_op_down(op)
+            if self.state == ReconnectState.WAITING_TO_RECONNECT:
+                logger.info(
+                    "{}({}): State is {}.  Adding to wait list".format(
+                        self.name, op.name, self.state
+                    )
+                )
+                self.waiting_connect_ops.append(op)
+            else:
+                logger.info(
+                    "{}({}): State is {}.  Adding to wait list and sending new connect op down".format(
+                        self.name, op.name, self.state
+                    )
+                )
+                self.waiting_connect_ops.append(op)
+                self._send_new_connect_op_down()
```

Comment:
I don't quite follow here.

So, some ConnectOperation comes down the pipe. We are NOT waiting to reconnect, so it is put in the list of operations that are waiting (?), and then a new ConnectOperation is sent down instead?